homeassistant/components/teslemetry/update.py

```diff
@@ -0,0 +1,104 @@
+"""Update platform for Teslemetry integration."""
+
+from __future__ import annotations
+
+from typing import Any, cast
+
+from tesla_fleet_api.const import Scope
+
+from homeassistant.components.update import UpdateEntity, UpdateEntityFeature
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .entity import TeslemetryVehicleEntity
+from .models import TeslemetryVehicleData
+
+AVAILABLE = "available"
+DOWNLOADING = "downloading"
+INSTALLING = "installing"
+WIFI_WAIT = "downloading_wifi_wait"
+SCHEDULED = "scheduled"
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the Teslemetry update platform from a config entry."""
+
+    async_add_entities(
+        TeslemetryUpdateEntity(vehicle, entry.runtime_data.scopes)
+        for vehicle in entry.runtime_data.vehicles
+    )
+
+
+class TeslemetryUpdateEntity(TeslemetryVehicleEntity, UpdateEntity):
+    """Teslemetry Updates entity."""
+
+    def __init__(
+        self,
+        data: TeslemetryVehicleData,
+        scopes: list[Scope],
+    ) -> None:
+        """Initialize the Update."""
+        self.scoped = Scope.VEHICLE_CMDS in scopes
+        super().__init__(
+            data,
+            "vehicle_state_software_update_status",
+        )
+
+    def _async_update_attrs(self) -> None:
+        """Update the attributes of the entity."""
+
+        # Supported Features
+        if self.scoped and self._value in (
+            AVAILABLE,
+            SCHEDULED,
+        ):
+            # Only allow install when an update is available
+            self._attr_supported_features = (
+                UpdateEntityFeature.PROGRESS | UpdateEntityFeature.INSTALL
+            )
+        else:
+            self._attr_supported_features = UpdateEntityFeature.PROGRESS
```

Comment:
```suggestion
```
This isn't correct. The flags should be removed if an update can't be installed for some reason in that case. They should not be removed just because there isn't an update available.

Reply:
The update can only be installed when the status is available or scheduled. The only difference between the code and what you describe is that yes, INSTALL is also removed when there is no update at all.

This implementation matches Tessie: https://github.com/home-assistant/core/blob/dev/homeassistant/components/tessie/update.py#L43

Reply:
Still seems a bit odd but I guess that makes sense somehow.

Reply:
The comment in the code was misleading, I will adjust that to be clearer that you can only install when downloading is finished.